File: SMObjectiveCSample/SMObjectiveCSample/Cells/InputCell.m
//
// Copyright (c) 2021 Soul Machines. All rights reserved.
//  InputCell.m
//  SMObjectiveCSample
//

#import "InputCell.h"

@interface InputCell () <UITextFieldDelegate>

@property ConfigId backingConfigId;

@end

@implementation InputCell

@synthesize backingConfigId;

+ (NSString *) identifier
{
    return NSStringFromClass(InputCell.self);
}

- (void) prepareForReuse
{
    [super prepareForReuse];
    [self.inputField setText:@""];
    [self.inputLabel setText:@""];
    
    [self setUserInteractionEnabled:true];
    [self.inputLabel setTextColor:UIColor.labelColor];
}

- (void) setConfigId:(ConfigId) configId
{
    self.backingConfigId = configId;
    NSString *defaultsIdentifier = [Enums labelFromConfigId:self.backingConfigId];
    [self.inputLabel setText:NSLocalizedString(defaultsIdentifier, nil)];

    NSString *textContent = [NSUserDefaults.standardUserDefaults stringForKey:defaultsIdentifier];
    [self.inputField setText:textContent];
}

- (void) textFieldDidEndEditing:(UITextField *)textField
{
    NSString *textContent = self.inputField.text;
    NSString *defaultsIdentifier = [Enums labelFromConfigId:self.backingConfigId];
    
    [NSUserDefaults.standardUserDefaults setObject:textContent forKey:defaultsIdentifier];
}

- (void) updateContent
{
    [super updateContent];
    [self.inputLabel setTextColor:([self isUserInteractionEnabled] ? UIColor.labelColor : UIColor.secondaryLabelColor)];
}

@end

File: SMSwiftSample/SMSwiftSample/Cells/InputCell.swift
//
// Copyright (c) 2022 Soul Machines. All rights reserved.
//  InputCell.swift
//  SMSwiftSample
//

import UIKit

class InputCell: SampleCustomCell {
    static let identifier = String(describing: InputCell.self)
    
    @IBOutlet private weak var inputField: UITextField?
    @IBOutlet private weak var inputLabel: UILabel?
    
    private var configId: ConfigId?
    
    override func prepareForReuse() {
        self.inputField?.text = ""
        self.inputLabel?.text = ""
        
        self.inputField?.textColor = UIColor.label
        self.isUserInteractionEnabled = true
    }
    
    func set(configId: ConfigId) {
        self.configId = configId
        self.inputLabel?.text = NSLocalizedString(configId.rawValue, comment: "")
        
        let textContent = UserDefaults.standard.string(forKey: configId.rawValue)
        self.inputField?.text = textContent
    }
    
    override func updateContent() {
        super.updateContent()
        inputLabel?.textColor = self.isUserInteractionEnabled ? .label : .secondaryLabel
    }
}

extension InputCell: UITextFieldDelegate {
    func textFieldDidEndEditing(_ textField: UITextField) {
        if let textContent = self.inputField?.text, let id = self.configId?.rawValue {
            UserDefaults.standard.set(textContent, forKey: id)
        }
    }
}
